ArduCopter/precision_landing.cpp

```diff
@@ -22,4 +27,197 @@ void Copter::update_precland()
 
     precland.update(height_above_ground_cm, rangefinder_alt_ok());
 }
+
+
+void Mode::PrecLand_StateMachine::init()
+{
+    // init is only called ONCE per mode change. So in a particuar mode we can retry only a finite times.
+    // The counter will be reset if the statemachine is called from a different mode
+    _retry_count = 0;
+    // reset every other statemachine
+    reset_failed_landing_statemachines();
+}
+
+// Reset the landing statemachines. This needs to be called everytime the landing target is back in sight.
+// So that if the landing target goes out of sight again, we can start the failed landing procedure back from the beginning stage
+void Mode::PrecLand_StateMachine::reset_failed_landing_statemachines()
+{
+    landing_target_lost_action = Landing_Target_Lost_Action::INIT;
+    retry_status = RetryLanding::RETRY_INIT;
+    failsafe_initialized = false;
+}
+
+// Run Prec Land State Machine. During Prec Landing, we might encouter four scenarios:
+// 1. We had the target in sight, but have lost it now. 2. We never had the target in sight and user wants to land.
+// 3. We have the target in sight and can continue landing. 4. The sensor is out of range
+// This method deals with all of these scenarios
+void Mode::PrecLand_StateMachine::update_precland_state_machine()
+{
+    // grab the current status of Landing Target
+    AC_PrecLand::PldState current_state =  copter.precland.get_plnd_target_status();
+
+    switch (current_state)
+    {
+    case AC_PrecLand::PldState::TARGET_RECENTLY_LOST:
+        // we have lost the target but had it in sight atleast once recently
+        // action will depend on what user wants
+        target_lost_actions();
+        break;
+
+    case AC_PrecLand::PldState::TARGET_LOST:
+        // we have no clue where we are supposed to be landing
+        // let user decide how strict our failsafe actions need to be
+        do_failsafe_actions();
+        break;
+
+    case AC_PrecLand::PldState::TARGET_OUT_OF_RANGE:
+        // The target isn't in sight, but we can't run any fail safe measures or do landing retry
+        // Therefore just descend for now, and check again later if retry is allowed
+    case AC_PrecLand::PldState::TARGET_FOUND:
+    default:
+        // no action required, target is in sight
+        copter.flightmode->run_land_controller();
+        reset_failed_landing_statemachines();
+        break;
+    }
+}
+
+
+// Target is lost (i.e we had it in sight some time back), this method helps decide on what needs to be done next
+// The chosen action depends on user set landing strictness
+void Mode::PrecLand_StateMachine::target_lost_actions()
+{
+    switch (landing_target_lost_action)
+    {
+    case Landing_Target_Lost_Action::INIT:
+    {
```

Comment:
```suggestion
    case TargetLostAction::INIT: {
```

Underscores are definitely not typical (similarly elsewhere).

`Landing` is kind of redundant in the enum name :-)